ambry-api/src/main/java/com/github/ambry/rest/RestUtils.java

```diff
@@ -1022,4 +1046,53 @@ public static String stripSlashAndExtensionFromId(String blobIdWithExtension) {
     int endIndex = extensionIndex != -1 ? extensionIndex : blobIdWithExtension.length();
     return blobIdWithExtension.substring(startIndex, endIndex);
   }
+
+  /**
+   * Class to encode decode kv header values.
+   */
+  public static class KVHeaderValueEncoderDecoder {
+    private static String DELIM = "; ";
+    private static String KV_SEPERATOR = "=";
+
+    /**
+     * Encode a map of {@link String}s as http header value.
+     * @param map {@link Map} of {@link String}s to encode.
+     * @return encoded http header value.
+     */
+    public static String encodeKVHeaderValue(Map<String, String> map) {
+      Objects.requireNonNull(map);
+      String value = "";
+      for (Map.Entry<String, String> entry : map.entrySet()) {
+        value += entry.getKey() + KV_SEPERATOR + entry.getValue() + DELIM;
+      }
+      return value;
+    }
+
+    /**
+     * Decode a kv header value.
+     * @param value {@link String} containing the encoded kv values.
+     * @return decoded kv {@link Map}.
+     */
+    public static Map<String, String> decodeKVHeaderValue(String value) {
+      Objects.requireNonNull(value);
+      Map<String, String> valueMap = new HashMap<>();
+      if (value.isEmpty()) {
+        return valueMap;
+      }
+      if (!value.contains(DELIM)) {
```

Comment:
All in stream ?
```
      return Arrays.stream(value.split(DELIM)).map(kv -> kv.split(KV_SEPERATOR)).peek(kv -> {
        if (kv.length != 2) {
          throw new IllegalArgumentException("Invalue kv header value: " + kv);
        }
      }).collect(Collectors.toMap(kv -> kv[0], kv -> kv[1]));```

Reply:
I find that hard to read, even if it works.